ui/src/helpers/vulnerability-statistics.ts

```diff
@@ -0,0 +1,276 @@
+export function findHighestCvssScore(
+  vulnerability: Vulnerability
+): CvssScoreResult | undefined {
+  const cvssReferences = extractCvssReferences(vulnerability);
+  if (cvssReferences.length === 0) {
+    return undefined;
+  }
+  const highestCvss = cvssReferences.reduce((prev, current) =>
+    prev.score > current.score ? prev : current
+  );
+
+  switch (highestCvss.version) {
+    case 'CVSS:2.0': {
+      const cvss = new Cvss2(highestCvss.vector);
+      return {
+        version: 'CVSS:2.0',
+        scores: cvss.calculateScores(true),
+      };
+    }
+    case 'CVSS:3.0': {
+      const cvss = new Cvss3P0(highestCvss.vector);
+      return {
+        version: 'CVSS:3.0',
+        scores: cvss.calculateScores(true),
+      };
+    }
+    case 'CVSS:3.1': {
+      const cvss = new Cvss3P1(highestCvss.vector);
+      return {
+        version: 'CVSS:3.1',
+        scores: cvss.calculateScores(true),
+      };
+    }
+    // CVSS4.0 gives the results in a different format than 2.0, 3.0 and 3.1, by using
+    // macrovectors to expose the different components of the CVSS vector.
+    case 'CVSS:4.0': {
+      const cvss = new Cvss4P0(highestCvss.vector);
+      return {
+        version: 'CVSS:4.0',
+        scores: cvss.calculateScores(true),
+        macroVector: {
+          name: cvss.getMacroVector().toString(),
+          exploitability: levelToText(
+            cvss.getMacroVector().getEq1().getLevelAsInt()
+          ),
+          complexity: levelToText(
+            cvss.getMacroVector().getEq2().getLevelAsInt()
+          ),
+          vulnerableSystem: levelToText(
+            cvss.getMacroVector().getEq3().getLevelAsInt()
+          ),
+          subsequentSystem: levelToText(
+            cvss.getMacroVector().getEq4().getLevelAsInt()
+          ),
+          exploitation: levelToText(
+            cvss.getMacroVector().getEq5().getLevelAsInt()
+          ),
+          sequrityRequirements: levelToText(
+            cvss.getMacroVector().getEq6().getLevelAsInt()
+          ),
+        },
+      };
+    }
+    default:
+      return undefined;
+  }
+}
+
+/**
+ * Extracts all valid EPSS scores and percentiles from the given vulnerability.
+ *
+ * @param vulnerability The vulnerability object containing references.
+ * @returns An array of objects containing the EPSS `score` and `percentile`.
+ */
+function extractEpssReferences(vulnerability: Vulnerability): EpssData[] {
+  // Filter out references that are not marked as EPSS or the score/vector is not a number,
+  // to avoid unnecessary processing.
+  const references = vulnerability.references.filter(
+    (ref) =>
+      ref.scoringSystem === 'epss' &&
+      !isNaN(Number(ref.score)) &&
+      !isNaN(Number(ref.vector))
+  );
+  const validReferences: EpssData[] = [];
+
+  for (const reference of references) {
+    const score = Number(reference.score);
+    const percentile = Number(reference.vector); // For some reason, ORT returns the percentile as the vector.
```

Comment:
Regarding the comment: I guess that happens only with the OSV advisor, which seems to [map the native `score` to ORT's `vector`](https://github.com/oss-review-toolkit/ort/blob/3239b719a6e800a18ba330d22264b9e3ccabbcab/plugins/advisors/osv/src/main/kotlin/Osv.kt#L149-L153).

Reply:
I just created https://github.com/oss-review-toolkit/ort/pull/10161 to clarify.